Support eq and notEq

herndlm opened the pull request:
I don't think they add much value, but why not.
Maybe that loosely type narrowing helps prevent errors somewhere :)

herndlm committed:
Support eq and notEq

ondrejmirtes commented:
Please add some tests that might trick the current implementation and the problems we're trying to solve, like $a == $a (always true) and also createStdClass() == createStdClass() (not always true). Thanks 😊

herndlm committed:
Add impossible-check tests for eq and notEq

herndlm commented:
added some more tests, but I'm not entirely sure what I'm finding here and what is expected and what not. I think this needs a double-check from you :)
`==` is quite tricky for me compared to `===` where things are pretty clear

And while looking at https://www.php.net/manual/en/language.operators.comparison.php I think we could come up with more weird constant comparisons where type-juggling is involved, like `100 == "1e2"` which always evaluates to equal, but I'm not sure if this is still relevant here.

herndlm commented:
Btw. I'm going to add more impossible-check tests for `===` in a similar fashion in another PR

ondrejmirtes commented on tests/Type/WebMozartAssert/data/impossible-check-eq-not-eq.php:
```diff
@@ -0,0 +1,39 @@
+function instancesOfDifferentTypesAreNeverEqual(stdClass $a, DateTimeInterface $b, stdClass $c, DateTimeInterface $d, stdClass $e, stdClass $f, stdClass $g, stdClass $h): void
+{
+	Assert::eq($a, $b); // ignored, should evaluate to false?
+	Assert::notEq($c, $d); // ignored, should evaluate to true?
```
These are fine. PHPStan currently doesn't implement the messy semantics from here: https://www.php.net/manual/en/types.comparisons.php ("Loose comparisons with ==" table).

herndlm replied:
so it just doesn't support `==` here at all, right? Also for instances of different classes?

ondrejmirtes replied:
Yes, it's not currently handled at all.

I just pushed some changes to phpstan/phpstan-src dev-master, I'm pretty sure it will change some results here after it lands in phpstan/phpstan (once this finishes https://github.com/phpstan/phpstan-src/actions/runs/1957754903). Feel free to wait a bit and require ^1.4.9 here.

herndlm replied:
great. Just for my understanding, is it easy to explain why the changes you made (e.g. https://github.com/phpstan/phpstan-src/commit/98c4d84ec983fb36992119d4c393d7192a2a99b6) are in MutatingScope and not in TypeSpecifier?

herndlm replied:
aah, I assume you're kind of building or simplifying expressions there for the TypeSpecifier to specify later. Makes sense then

ondrejmirtes replied:
That commit isn't related to TypeSpecifier as you can see in the attached tests :) But I might still use it if I decide to somehow work with `$scope->getType($equal)` in the TypeSpecifier to fix the regressions in my PR https://github.com/phpstan/phpstan-src/pull/1058.

ondrejmirtes commented on tests/Type/WebMozartAssert/data/impossible-check-eq-not-eq.php:
```diff
@@ -0,0 +1,39 @@
+	Assert::eq($a, $b); // ignored, should evaluate to false?
+	Assert::notEq($c, $d); // ignored, should evaluate to true?
+
+	Assert::eq($e, new stdClass()); // will always evaluate to true, should evaluate to false?
```
Should not report anything

ondrejmirtes commented on tests/Type/WebMozartAssert/data/impossible-check-eq-not-eq.php:
```diff
@@ -0,0 +1,39 @@
+	Assert::notEq($c, $d); // ignored, should evaluate to true?
+
+	Assert::eq($e, new stdClass()); // will always evaluate to true, should evaluate to false?
+	Assert::notEq($f, new stdClass()); // ignored, should evaluate to true?
```
Should not report anything

herndlm committed:
Update tests with info received in review, rename test method to match the filename

ondrejmirtes commented on tests/Type/WebMozartAssert/data/impossible-check-eq-not-eq.php:
```diff
@@ -0,0 +1,40 @@
+	Assert::eq($a, $b);
+	Assert::notEq($c, $d);
+
+	Assert::eq($e, new stdClass()); // will always evaluate to true, should not report anything
```
Please adjust the asserts to what actually makes sense, and not what PHPStan currently reports. I'm pretty sure this one "'Call to static method Webmozart\Assert\Assert::eq() with stdClass and stdClass will always evaluate to true.'" on line 30 doesn't make sense. There might be more like that.

herndlm replied:
adapted, sorry it is getting messy.. interestingly most of the loose scalar comparison seem to be working fine as well, except '1' == 1. And also weird: nothing seems to have changed with the latest phpstan commits hmm

ondrejmirtes replied:
No idea why some errors `'Call to static method Webmozart\Assert\Assert::notEq() with 1 and true will always evaluate to false.'` actually work. It'd be great to find out which line in TypeSpecifier is responsible for that.

herndlm replied:
Must be the parts that continue specification if left and right expr strings are different or there are variables involved? https://github.com/phpstan/phpstan-src/blob/master/src/Analyser/TypeSpecifier.php#L452 But I'm just guessing now tbh.

herndlm committed:
More impossible-check eq and notEq tests, target next PHPStan version

ondrejmirtes committed:
Commented out what currently PHPStan does not detect

I'm surprised that "with 1 and true will always evaluate to true"
is even detected.

ondrejmirtes commented:
There's still one failure and I have no idea why.

ondrejmirtes commented:
Alright ,the error "+54: Call to static method Webmozart\Assert\Assert::eq() with stdClass and mixed will always evaluate to true." should not be reported, but the root issue here is that PHPStan actually doesn't see the function declaration, because it's not loaded.

Please rewrite these tests to use classes+methods, it's more reliable (and you don't have to `require_once` those files).

herndlm committed:
Move impossible-check tests into class

ondrejmirtes commented:
Thank you!
